CheckedPtr should require FastMalloc, Part 2

geoffreygaren opened the pull request:
#### 8b7d41505065a3dacb612e0dc4112df118590c05
<pre>
CheckedPtr should require FastMalloc, Part 2
<a href="https://bugs.webkit.org/show_bug.cgi?id=271993">https://bugs.webkit.org/show_bug.cgi?id=271993</a>
<a href="https://rdar.apple.com/125737157">rdar://125737157</a>

Reviewed by Ryosuke Niwa.

This will enable zombie-mode CheckedPtr.

It&apos;s also just good performance hygiene.

The two major changes in this patch are:

(1) Declare your type when you inherit from CanMakeCheckedPtr, like we do for
RefCounted and CanMakeWeakPtr. We use this to do our FastMalloc static_assert.
(In future, I also plan to use this type to share our implementation with other
lifetime schemes.) Fixed any classes that violated the FastMalloc static_assert.

(2) Moved CanMakeCheckedPtr down from abstract base classes into concrete
subclasses. It&apos;s not possible for abstract base classes to require or
demonstrate that a class uses FastMalloc, or has any other concrete allocation
or destruction behavior. Abstract base classes that require a CheckedPtr
interface now make that part of their abstract API. (In future we need to do
this for CanMakeWeakPtr too. We&apos;ve already done it for RefCounted.) It&apos;s a bit
more obvious now which classes have overlapping / wasteful lifetime
implementations. We can clean that up soon.

* Source/WTF/wtf/CheckedRef.h:

(WTF::CanMakeCheckedPtrBase::~CanMakeCheckedPtrBase): Moved the FastMalloc
assertion here because CheckedPtr needs to support abstract base classes that
don&apos;t know how they&apos;re allocated. This is logically clearer: CheckedPtr just
invokes your checking API and doesn&apos;t care how you check; only CanMakeCheckedPtr
will depend on FastMalloc.

* Source/WebCore/Modules/speech/SpeechRecognitionCaptureSourceImpl.h: (2)

* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/animation/DocumentTimelinesController.h:
* Source/WebCore/bindings/js/ScriptController.h:
* Source/WebCore/dom/CustomElementDefaultARIA.h:
* Source/WebCore/dom/CustomElementReactionQueue.h:
* Source/WebCore/dom/DocumentMarkerController.h:
* Source/WebCore/dom/EventLoop.h:
* Source/WebCore/dom/ExtensionStyleSheets.h:
* Source/WebCore/dom/FullscreenManager.h:
* Source/WebCore/dom/IdTargetObserver.h:
* Source/WebCore/dom/IdTargetObserverRegistry.h: (1)

* Source/WebCore/dom/PendingScriptClient.h: (2)
* Source/WebCore/dom/RejectedPromiseTracker.h: (1)
* Source/WebCore/dom/ScriptRunner.h: (2)

* Source/WebCore/dom/VisitedLinkState.h:
* Source/WebCore/dom/messageports/MessagePortChannelRegistry.h:
* Source/WebCore/dom/messageports/WorkerMessagePortChannelProvider.h:
* Source/WebCore/editing/Editor.h:
* Source/WebCore/editing/FrameSelection.h:
* Source/WebCore/editing/TextManipulationController.h:
* Source/WebCore/history/BackForwardController.h:
* Source/WebCore/history/CachedPage.h: (1)

* Source/WebCore/html/parser/HTMLDocumentParser.h: (2)

* Source/WebCore/layout/integration/flex/LayoutIntegrationFlexLayout.h:
* Source/WebCore/layout/integration/inline/LayoutIntegrationLineLayout.h:
* Source/WebCore/layout/layouttree/LayoutBox.h:
(WebCore::Layout::Box::incrementPtrCount const): Deleted.
(WebCore::Layout::Box::decrementPtrCount const): Deleted.
* Source/WebCore/loader/FrameLoader.cpp:
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/loader/HistoryController.h:
* Source/WebCore/loader/NavigationScheduler.h:
* Source/WebCore/loader/ProgressTracker.h:
* Source/WebCore/loader/ThreadableLoaderClient.h:
* Source/WebCore/page/AlternativeTextClient.h:
* Source/WebCore/page/DeviceController.h:
* Source/WebCore/page/DiagnosticLoggingClient.h:
* Source/WebCore/page/EditorClient.h:
* Source/WebCore/page/ElementTargetingController.h:
* Source/WebCore/page/EventHandler.h:
* Source/WebCore/page/FocusController.h:
* Source/WebCore/page/LocalFrameViewLayoutContext.h:
* Source/WebCore/page/PageConsoleClient.h:
* Source/WebCore/page/csp/ContentSecurityPolicy.h:
* Source/WebCore/page/scrolling/ScrollingStateTree.h:
* Source/WebCore/platform/KeyboardScrollingAnimator.h:
* Source/WebCore/platform/OrientationNotifier.h:
* Source/WebCore/platform/ScrollView.h:
* Source/WebCore/platform/ScrollableArea.cpp:
* Source/WebCore/platform/ScrollableArea.h: (1)

* Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
* Source/WebCore/platform/cocoa/VideoPresentationModel.h:
* Source/WebCore/platform/cocoa/WebAVPlayerLayer.mm: (2)

* Source/WebCore/platform/graphics/DisplayRefreshMonitorClient.h: (1)

* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::FontCascade):
(WebCore::GlyphToPathTranslator::containsMorePaths):
(WebCore::GlyphToPathTranslator::path):
(WebCore::GlyphToPathTranslator::extents):
(WebCore::GlyphToPathTranslator::advance):
* Source/WebCore/platform/graphics/FontCascade.h:
* Source/WebCore/platform/graphics/GlyphBuffer.h: Removed the recently added ability to
use CheckedPtr&lt;GlyphBuffer&gt;. GlyphBuffer is stack-allocated so it can&apos;t
participate in a zombie-mode lifetime scheme. We&apos;ll have to solve GlyphBuffer
another way later. (Chris said he&apos;s OK with this.)

* Source/WebCore/platform/graphics/TextRun.cpp:
* Source/WebCore/platform/graphics/TextRun.h:
* Source/WebCore/platform/graphics/TiledBacking.h: (1)

* Source/WebCore/platform/graphics/WidthIterator.cpp:
(WebCore::WidthIterator::commitCurrentFontRange): See GlyphBuffer.h

* Source/WebCore/platform/graphics/ca/LayerPool.h: (1)
* Source/WebCore/platform/graphics/cocoa/NullPlaybackSessionInterface.h:
* Source/WebCore/platform/graphics/cocoa/NullVideoPresentationInterface.h:
* Source/WebCore/platform/ios/PlaybackSessionInterfaceIOS.h:
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.h:
* Source/WebCore/platform/mac/PlaybackSessionInterfaceMac.h:
* Source/WebCore/platform/mac/VideoPresentationInterfaceMac.h:
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivate.h:
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateAVFImpl.h:
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateMock.h:
* Source/WebCore/platform/mediastream/AudioTrackPrivateMediaStream.h:
* Source/WebCore/platform/mediastream/RealtimeMediaSource.h:
* Source/WebCore/platform/mediastream/mac/MediaStreamTrackAudioSourceProviderCocoa.h:
* Source/WebCore/platform/mediastream/mac/RealtimeOutgoingAudioSourceCocoa.h: (2)

* Source/WebCore/platform/network/curl/CurlMultipartHandleClient.h:
* Source/WebCore/platform/sql/SQLiteDatabase.h:
* Source/WebCore/rendering/LegacyRootInlineBox.cpp:
* Source/WebCore/rendering/LegacyRootInlineBox.h:
* Source/WebCore/rendering/MarkedText.h:
* Source/WebCore/rendering/RegionContext.h:
* Source/WebCore/rendering/RenderImageResource.h:
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderObject.cpp:
* Source/WebCore/rendering/RenderObject.h:
* Source/WebCore/style/StyleScope.h:
* Source/WebCore/style/StyleUpdate.h:
* Source/WebCore/svg/SVGDocumentExtensions.h: (1)

* Source/WebCore/testing/Internals.h: (2)
* Source/WebCore/workers/WorkerOrWorkletScriptController.h:
* Source/WebCore/workers/service/server/SWServer.h:
* Source/WebCore/workers/service/server/SWServerJobQueue.h:
* Source/WebCore/workers/service/server/SWServerToContextConnection.h: (1)

* Source/WebCore/xml/parser/XMLDocumentParser.h:
* Source/WebKit/NetworkProcess/Downloads/DownloadManager.h:
* Source/WebKit/NetworkProcess/NetworkProcess.h: (2)

* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerNavigationPreloader.h:
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerSoftUpdateLoader.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebPreferencesStore.h:
(WebKit::WebPreferencesStore::isolatedCopy const):
(WebKit::WebPreferencesStore::isolatedCopy):
* Source/WebKit/Shared/WebPreferencesStore.serialization.in:
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
* Source/WebKit/UIProcess/Cocoa/ExtensionCapabilityGranter.h: (1)

* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp: (2)

* Source/WebKit/UIProcess/DisplayLink.h:
* Source/WebKit/UIProcess/Inspector/RemoteWebInspectorUIProxy.h:
* Source/WebKit/UIProcess/ProcessThrottler.cpp:
* Source/WebKit/UIProcess/SuspendedPageProxy.h:
* Source/WebKit/UIProcess/WebBackForwardCache.h:
* Source/WebKit/UIProcess/WebProcessCache.h: (1)

* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::establishRemoteWorkerContextConnectionToNetworkProcess):
WebPreferencesStore&amp; can&apos;t use zombie-mode CheckedPtr because it&apos;s an interior
object pointer. For now, we can accuratrely represent lifetime by explicitly
ref&apos;ing its owner. Later, we should just change it to be refcounted.

* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/WebProcess/Speech/SpeechRecognitionRealtimeMediaSourceManager.cpp:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.h:
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.h: (2)

* Source/WebKitLegacy/Storage/StorageThread.h:
* Source/WebKitLegacy/WebCoreSupport/WebResourceLoadScheduler.h:
* Tools/TestWebKitAPI/Tests/WTF/CheckedPtr.cpp:
* Tools/TestWebKitAPI/Tests/WTF/CheckedRef.cpp:
* Tools/TestWebKitAPI/Tests/WTF/Hasher.cpp:
(TestWebKitAPI::TEST(WTF, Hasher_CheckedPtr)): (1)

Canonical link: <a href="https://commits.webkit.org/277020@main">https://commits.webkit.org/277020@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/57356d21c37f6e0f6249b177c21ca08135fd5cc1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46412 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25556 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49008 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49087 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42453 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48719 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29917 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23046 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37884 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46990 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22596 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40026 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19125 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/19960 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41127 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4457 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/39654 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/42735 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41493 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50918 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/45889 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/21417 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/17891 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45121 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22708 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44052 "Passed tests") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/23097 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/53037 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6479 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22407 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10870 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/Modules/speech/SpeechRecognitionCaptureSourceImpl.h:
```diff
@@ -58,6 +59,11 @@ class SpeechRecognitionCaptureSourceImpl
 
 private:
+    // CheckedPtr interface
+    uint32_t ptrCount() const final { return CanMakeCheckedPtr<SpeechRecognitionCaptureSourceImpl>::ptrCount(); }
```
Do we really need to specify SpeechRecognitionCaptureSourceImpl here?

geoffreygaren replied:
I was able to remove the <SpeechRecognitionCaptureSourceImpl> specification by making CanMakeCheckedPtr a class instead of a typedef. This might reduce template bloat too. Thanks for the suggestion!

rniwa commented on Source/WebCore/dom/ScriptRunner.h:
```diff
@@ -39,12 +40,17 @@ class ScriptElement;
     ~ScriptRunner();
 
+    // CheckedPtr interface
+    uint32_t ptrCount() const final { return CanMakeCheckedPtr<ScriptRunner>::ptrCount(); }
```
Ditto about specifying ScriptRunner.

rniwa commented on Source/WebCore/html/parser/HTMLDocumentParser.h:
```diff
@@ -45,12 +46,17 @@ class HTMLResourcePreloader;
     virtual ~HTMLDocumentParser();
 
+    // CheckedPtr interface
+    uint32_t ptrCount() const final { return CanMakeCheckedPtr<HTMLDocumentParser>::ptrCount(); }
```
Ditto.

rniwa commented on Source/WebCore/platform/cocoa/WebAVPlayerLayer.mm:
```diff
@@ -57,14 +57,19 @@ @interface WebAVPlayerLayer (Logging)
     }
 private:
+    // CheckedPtr interface
+    uint32_t ptrCount() const final { return CanMakeCheckedPtr<WebAVPlayerLayerPresentationModelClient>::ptrCount(); }
```
Ditto.

rniwa commented on Source/WebCore/platform/graphics/cocoa/NullPlaybackSessionInterface.h:
```diff
@@ -72,6 +74,11 @@ class NullPlaybackSessionInterface final
     }
 
+    // CheckedPtr interface
+    uint32_t ptrCount() const final { return CanMakeCheckedPtr<NullPlaybackSessionInterface>::ptrCount(); }
```
Ditto about specifying `NullPlaybackSessionInterface`.

rniwa commented on Source/WebCore/platform/graphics/cocoa/NullVideoPresentationInterface.h:
```diff
@@ -87,6 +88,11 @@ class NullVideoPresentationInterface final
     }
 
+    // CheckedPtr interface
+    uint32_t ptrCount() const final { return CanMakeCheckedPtr<NullVideoPresentationInterface>::ptrCount(); }
```
Ditto.

rniwa commented on Source/WebCore/platform/ios/PlaybackSessionInterfaceIOS.mm:
```diff
@@ -119,6 +119,22 @@
+
+uint32_t PlaybackSessionInterfaceIOS::ptrCount() const
+{
+    return CanMakeCheckedPtr<PlaybackSessionInterfaceIOS>::ptrCount();
```
Ditto. Also, why aren't these inlined like other functions?

geoffreygaren replied:
Style checker complains when you inline a virtual function in an exported class. Probably because it makes linkage weird? I didn't think too much about it. Just obeyed the bot.

rniwa commented on Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.h:
```diff
@@ -229,6 +230,11 @@ class VideoPresentationInterfaceIOS
 #endif
 private:
+    // CheckedPtr interface
+    uint32_t ptrCount() const final { return CanMakeCheckedPtr<VideoPresentationInterfaceIOS>::ptrCount(); }
```
Ditto. Do we need to specify VideoPresentationInterfaceIOS here?

rniwa commented on Source/WebCore/platform/ios/WebVideoFullscreenControllerAVKit.mm:
```diff
@@ -119,6 +121,11 @@ -(void)didFinishFullscreen:(VideoFullscreenControllerContext*)context;
     VideoFullscreenControllerContext() { }
 
+    // CheckedPtr interface
+    uint32_t ptrCount() const final { return CanMakeCheckedPtr<VideoFullscreenControllerContext>::ptrCount(); }
```
Ditto.

rniwa commented on Source/WebCore/platform/mac/PlaybackSessionInterfaceMac.mm:
```diff
@@ -294,6 +294,21 @@
 
+uint32_t PlaybackSessionInterfaceMac::ptrCount() const
+{
+    return CanMakeCheckedPtr<PlaybackSessionInterfaceMac>::ptrCount();
```
Ditto. Also why not inline these?

rniwa commented on Source/WebCore/platform/mac/VideoPresentationInterfaceMac.h:
```diff
@@ -110,6 +113,12 @@ class VideoPresentationInterfaceMac
     WEBCORE_EXPORT VideoPresentationInterfaceMac(PlaybackSessionInterfaceMac&);
+
+    // CheckedPtr interface
+    uint32_t ptrCount() const final { return CanMakeCheckedPtr<VideoPresentationInterfaceMac>::ptrCount(); }
```
Ditto.

rniwa commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivate.h:
```diff
@@ -89,6 +92,11 @@ class MediaRecorderPrivate
 
 private:
+    // CheckedPtr interface
+    uint32_t ptrCount() const final { return CanMakeCheckedPtr<MediaRecorderPrivate>::ptrCount(); }
```
Ditto.

rniwa commented on Source/WebCore/platform/mediastream/AudioTrackPrivateMediaStream.h:
```diff
@@ -71,6 +73,11 @@ class AudioTrackPrivateMediaStream final
     explicit AudioTrackPrivateMediaStream(MediaStreamTrackPrivate&);
 
+    // CheckedPtr interface
+    uint32_t ptrCount() const final { return CanMakeCheckedPtr<AudioTrackPrivateMediaStream>::ptrCount(); }
```
Ditto.

rniwa commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerMediaStreamSource.cpp:
```diff
@@ -522,6 +524,11 @@ class InternalSource final : public MediaStreamTrackPrivate::Observer,
 
 private:
+    // CheckedPtr interface
+    uint32_t ptrCount() const final { return CanMakeCheckedPtr<InternalSource>::ptrCount(); }
```
Ditto.

rniwa commented on Source/WebCore/platform/mediastream/mac/MediaStreamTrackAudioSourceProviderCocoa.h:
```diff
@@ -30,21 +30,28 @@
     explicit MediaStreamTrackAudioSourceProviderCocoa(MediaStreamTrackPrivate&);
 
+    // CheckedPtr interface
+    uint32_t ptrCount() const final { return CanMakeCheckedPtr<MediaStreamTrackAudioSourceProviderCocoa>::ptrCount(); }
```
Ditto.

rniwa commented on Source/WebCore/platform/mediastream/mac/RealtimeOutgoingAudioSourceCocoa.h:
```diff
@@ -37,14 +38,19 @@ class AudioTrackSinkInterface;
     ~RealtimeOutgoingAudioSourceCocoa();
 
+    // CheckedPtr interface
+    uint32_t ptrCount() const final { return CanMakeCheckedPtr<RealtimeOutgoingAudioSourceCocoa>::ptrCount(); }
```
Ditto.

rniwa commented on Source/WebCore/platform/network/curl/CurlRequest.h:
```diff
@@ -82,6 +82,11 @@ class CurlRequest : public ThreadSafeRefCounted<CurlRequest>, public CurlRequest
     WEBCORE_EXPORT CurlRequest(const ResourceRequest&, CurlRequestClient*, CaptureNetworkLoadMetrics);
 
+    // CheckedPtr interface
+    uint32_t ptrCount() const final { return CanMakeThreadSafeCheckedPtr<CurlRequest>::ptrCount(); }
```
Ditto.

rniwa commented on Source/WebCore/testing/Internals.h:
```diff
@@ -1459,6 +1461,14 @@ class Internals final : public RefCounted<Internals>, private ContextDestruction
+
+#if ENABLE(MEDIA_STREAM)
+    // CheckedPtr interface
+    uint32_t ptrCount() const final { return CanMakeCheckedPtr<Internals>::ptrCount(); }
```
Ditto.

rniwa commented on Source/WebCore/xml/parser/XMLDocumentParser.h:
```diff
@@ -91,6 +92,11 @@ class XMLDocumentParser final : public ScriptableDocumentParser, public PendingS
     XMLDocumentParser(DocumentFragment&, HashMap<AtomString, AtomString>&&, const AtomString&, OptionSet<ParserContentPolicy>);
 
+    // CheckedPtr interface
+    uint32_t ptrCount() const final { return CanMakeCheckedPtr<XMLDocumentParser>::ptrCount(); }
```
Ditto.

rniwa commented on Source/WebKit/NetworkProcess/NetworkProcess.h:
```diff
@@ -430,6 +431,11 @@ class NetworkProcess : public AuxiliaryProcess, private DownloadManager::Client,
 
 private:
+    // CheckedPtr interface
+    uint32_t ptrCount() const final { return CanMakeCheckedPtr<NetworkProcess>::ptrCount(); }
```
Ditto.

rniwa commented on Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
```diff
@@ -271,6 +273,12 @@ class UserMediaCaptureManagerProxy::SourceProxy
 private:
+
+    // CheckedPtr interface
+    uint32_t ptrCount() const final { return CanMakeCheckedPtr<UserMediaCaptureManagerProxy::SourceProxy>::ptrCount(); }
```
Ditto.

rniwa commented on Source/WebKit/WebProcess/Speech/SpeechRecognitionRealtimeMediaSourceManager.cpp:
```diff
@@ -85,6 +87,11 @@ class SpeechRecognitionRealtimeMediaSourceManager::Source
 
 private:
+    // CheckedPtr interface
+    uint32_t ptrCount() const final { return CanMakeCheckedPtr<SpeechRecognitionRealtimeMediaSourceManager::Source>::ptrCount(); }
```
Ditto.

rniwa commented on Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.h:
```diff
@@ -70,6 +72,11 @@ class PlaybackSessionInterfaceContext final
     friend class VideoPresentationInterfaceContext;
 
+    // CheckedPtr interface
+    uint32_t ptrCount() const final { return CanMakeCheckedPtr<PlaybackSessionInterfaceContext>::ptrCount(); }
```
Ditto.

rniwa commented on Source/WebKit/WebProcess/cocoa/VideoPresentationManager.h:
```diff
@@ -101,6 +103,12 @@ class VideoPresentationInterfaceContext
     void hasVideoChanged(bool) override;
+
+    // CheckedPtr interface
+    uint32_t ptrCount() const final { return CanMakeCheckedPtr<VideoPresentationInterfaceContext>::ptrCount(); }
```
Ditto.

rniwa commented on Tools/TestWebKitAPI/Tests/WebCore/curl/CurlMultipartHandleTests.cpp:
```diff
@@ -91,6 +92,11 @@ class MultipartHandleClient : public CurlMultipartHandleClient {
 
 private:
+    // CheckedPtr interface
+    uint32_t ptrCount() const final { return CanMakeCheckedPtr<MultipartHandleClient>::ptrCount(); }
```
Ditto.

geoffreygaren committed:
CheckedPtr should require FastMalloc, Part 2
https://bugs.webkit.org/show_bug.cgi?id=271993
rdar://125737157

Reviewed by Ryosuke Niwa.

This will enable zombie-mode CheckedPtr.

It's also just good performance hygiene.

The two major changes in this patch are:

(1) Declare your type when you inherit from CanMakeCheckedPtr, like we do for
RefCounted and CanMakeWeakPtr. We use this to do our FastMalloc static_assert.
(In future, I also plan to use this type to share our implementation with other
lifetime schemes.) Fixed any classes that violated the FastMalloc static_assert.

(2) Moved CanMakeCheckedPtr down from abstract base classes into concrete
subclasses. It's not possible for abstract base classes to require or
demonstrate that a class uses FastMalloc, or has any other concrete allocation
or destruction behavior. Abstract base classes that require a CheckedPtr
interface now make that part of their abstract API. (In future we need to do
this for CanMakeWeakPtr too. We've already done it for RefCounted.) It's a bit
more obvious now which classes have overlapping / wasteful lifetime
implementations. We can clean that up soon.

* Source/WTF/wtf/CheckedRef.h:

(WTF::CanMakeCheckedPtrBase::~CanMakeCheckedPtrBase): Moved the FastMalloc
assertion here because CheckedPtr needs to support abstract base classes that
don't know how they're allocated. This is logically clearer: CheckedPtr just
invokes your checking API and doesn't care how you check; only CanMakeCheckedPtr
will depend on FastMalloc.

* Source/WebCore/Modules/speech/SpeechRecognitionCaptureSourceImpl.h: (2)

* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/animation/DocumentTimelinesController.h:
* Source/WebCore/bindings/js/ScriptController.h:
* Source/WebCore/dom/CustomElementDefaultARIA.h:
* Source/WebCore/dom/CustomElementReactionQueue.h:
* Source/WebCore/dom/DocumentMarkerController.h:
* Source/WebCore/dom/EventLoop.h:
* Source/WebCore/dom/ExtensionStyleSheets.h:
* Source/WebCore/dom/FullscreenManager.h:
* Source/WebCore/dom/IdTargetObserver.h:
* Source/WebCore/dom/IdTargetObserverRegistry.h: (1)

* Source/WebCore/dom/PendingScriptClient.h: (2)
* Source/WebCore/dom/RejectedPromiseTracker.h: (1)
* Source/WebCore/dom/ScriptRunner.h: (2)

* Source/WebCore/dom/VisitedLinkState.h:
* Source/WebCore/dom/messageports/MessagePortChannelRegistry.h:
* Source/WebCore/dom/messageports/WorkerMessagePortChannelProvider.h:
* Source/WebCore/editing/Editor.h:
* Source/WebCore/editing/FrameSelection.h:
* Source/WebCore/editing/TextManipulationController.h:
* Source/WebCore/history/BackForwardController.h:
* Source/WebCore/history/CachedPage.h: (1)

* Source/WebCore/html/parser/HTMLDocumentParser.h: (2)

* Source/WebCore/layout/integration/flex/LayoutIntegrationFlexLayout.h:
* Source/WebCore/layout/integration/inline/LayoutIntegrationLineLayout.h:
* Source/WebCore/layout/layouttree/LayoutBox.h:
(WebCore::Layout::Box::incrementPtrCount const): Deleted.
(WebCore::Layout::Box::decrementPtrCount const): Deleted.
* Source/WebCore/loader/FrameLoader.cpp:
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/loader/HistoryController.h:
* Source/WebCore/loader/NavigationScheduler.h:
* Source/WebCore/loader/ProgressTracker.h:
* Source/WebCore/loader/ThreadableLoaderClient.h:
* Source/WebCore/page/AlternativeTextClient.h:
* Source/WebCore/page/DeviceController.h:
* Source/WebCore/page/DiagnosticLoggingClient.h:
* Source/WebCore/page/EditorClient.h:
* Source/WebCore/page/ElementTargetingController.h:
* Source/WebCore/page/EventHandler.h:
* Source/WebCore/page/FocusController.h:
* Source/WebCore/page/LocalFrameViewLayoutContext.h:
* Source/WebCore/page/PageConsoleClient.h:
* Source/WebCore/page/csp/ContentSecurityPolicy.h:
* Source/WebCore/page/scrolling/ScrollingStateTree.h:
* Source/WebCore/platform/KeyboardScrollingAnimator.h:
* Source/WebCore/platform/OrientationNotifier.h:
* Source/WebCore/platform/ScrollView.h:
* Source/WebCore/platform/ScrollableArea.cpp:
* Source/WebCore/platform/ScrollableArea.h: (1)

* Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
* Source/WebCore/platform/cocoa/VideoPresentationModel.h:
* Source/WebCore/platform/cocoa/WebAVPlayerLayer.mm: (2)

* Source/WebCore/platform/graphics/DisplayRefreshMonitorClient.h: (1)

* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::FontCascade):
(WebCore::GlyphToPathTranslator::containsMorePaths):
(WebCore::GlyphToPathTranslator::path):
(WebCore::GlyphToPathTranslator::extents):
(WebCore::GlyphToPathTranslator::advance):
* Source/WebCore/platform/graphics/FontCascade.h:
* Source/WebCore/platform/graphics/GlyphBuffer.h: Removed the recently added ability to
use CheckedPtr<GlyphBuffer>. GlyphBuffer is stack-allocated so it can't
participate in a zombie-mode lifetime scheme. We'll have to solve GlyphBuffer
another way later. (Chris said he's OK with this.)

* Source/WebCore/platform/graphics/TextRun.cpp:
* Source/WebCore/platform/graphics/TextRun.h:
* Source/WebCore/platform/graphics/TiledBacking.h: (1)

* Source/WebCore/platform/graphics/WidthIterator.cpp:
(WebCore::WidthIterator::commitCurrentFontRange): See GlyphBuffer.h

* Source/WebCore/platform/graphics/ca/LayerPool.h: (1)
* Source/WebCore/platform/graphics/cocoa/NullPlaybackSessionInterface.h:
* Source/WebCore/platform/graphics/cocoa/NullVideoPresentationInterface.h:
* Source/WebCore/platform/ios/PlaybackSessionInterfaceIOS.h:
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.h:
* Source/WebCore/platform/mac/PlaybackSessionInterfaceMac.h:
* Source/WebCore/platform/mac/VideoPresentationInterfaceMac.h:
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivate.h:
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateAVFImpl.h:
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateMock.h:
* Source/WebCore/platform/mediastream/AudioTrackPrivateMediaStream.h:
* Source/WebCore/platform/mediastream/RealtimeMediaSource.h:
* Source/WebCore/platform/mediastream/mac/MediaStreamTrackAudioSourceProviderCocoa.h:
* Source/WebCore/platform/mediastream/mac/RealtimeOutgoingAudioSourceCocoa.h: (2)

* Source/WebCore/platform/network/curl/CurlMultipartHandleClient.h:
* Source/WebCore/platform/sql/SQLiteDatabase.h:
* Source/WebCore/rendering/LegacyRootInlineBox.cpp:
* Source/WebCore/rendering/LegacyRootInlineBox.h:
* Source/WebCore/rendering/MarkedText.h:
* Source/WebCore/rendering/RegionContext.h:
* Source/WebCore/rendering/RenderImageResource.h:
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderObject.cpp:
* Source/WebCore/rendering/RenderObject.h:
* Source/WebCore/style/StyleScope.h:
* Source/WebCore/style/StyleUpdate.h:
* Source/WebCore/svg/SVGDocumentExtensions.h: (1)

* Source/WebCore/testing/Internals.h: (2)
* Source/WebCore/workers/WorkerOrWorkletScriptController.h:
* Source/WebCore/workers/service/server/SWServer.h:
* Source/WebCore/workers/service/server/SWServerJobQueue.h:
* Source/WebCore/workers/service/server/SWServerToContextConnection.h: (1)

* Source/WebCore/xml/parser/XMLDocumentParser.h:
* Source/WebKit/NetworkProcess/Downloads/DownloadManager.h:
* Source/WebKit/NetworkProcess/NetworkProcess.h: (2)

* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerNavigationPreloader.h:
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerSoftUpdateLoader.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebPreferencesStore.h:
(WebKit::WebPreferencesStore::isolatedCopy const):
(WebKit::WebPreferencesStore::isolatedCopy):
* Source/WebKit/Shared/WebPreferencesStore.serialization.in:
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
* Source/WebKit/UIProcess/Cocoa/ExtensionCapabilityGranter.h: (1)

* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp: (2)

* Source/WebKit/UIProcess/DisplayLink.h:
* Source/WebKit/UIProcess/Inspector/RemoteWebInspectorUIProxy.h:
* Source/WebKit/UIProcess/ProcessThrottler.cpp:
* Source/WebKit/UIProcess/SuspendedPageProxy.h:
* Source/WebKit/UIProcess/WebBackForwardCache.h:
* Source/WebKit/UIProcess/WebProcessCache.h: (1)

* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::establishRemoteWorkerContextConnectionToNetworkProcess):
WebPreferencesStore& can't use zombie-mode CheckedPtr because it's an interior
object pointer. For now, we can accuratrely represent lifetime by explicitly
ref'ing its owner. Later, we should just change it to be refcounted.

* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/WebProcess/Speech/SpeechRecognitionRealtimeMediaSourceManager.cpp:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.h:
* Source/WebKit/WebProcess/cocoa/VideoPresentationManager.h: (2)

* Source/WebKitLegacy/Storage/StorageThread.h:
* Source/WebKitLegacy/WebCoreSupport/WebResourceLoadScheduler.h:
* Tools/TestWebKitAPI/Tests/WTF/CheckedPtr.cpp:
* Tools/TestWebKitAPI/Tests/WTF/CheckedRef.cpp:
* Tools/TestWebKitAPI/Tests/WTF/Hasher.cpp:
(TestWebKitAPI::TEST(WTF, Hasher_CheckedPtr)): (1)

Canonical link: https://commits.webkit.org/277020@main
